x-pack/plugins/fleet/server/routes/uninstall_token/handlers.ts

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { TypeOf } from '@kbn/config-schema';
+
+import type { PolicyUninstallTokenMap } from '../../../common/types/models/uninstall_token';
+
+import { appContextService } from '../../services';
+import type { UninstallTokenServiceInterface } from '../../services/security/uninstall_token_service';
+import type { FleetRequestHandler } from '../../types';
+import type { GetUninstallTokensResponse } from '../../../common/types/rest_spec/uninstall_token';
+import type { GetUninstallTokensRequestSchema } from '../../types/rest_spec/uninstall_token';
+
+export const getUninstallTokensHandler: FleetRequestHandler<
+  unknown,
+  TypeOf<typeof GetUninstallTokensRequestSchema.query>
+> = async (context, request, response) => {
+  const { agentTamperProtectionEnabled } = appContextService.getExperimentalFeatures();
+  if (!agentTamperProtectionEnabled) {
+    return response.customError({
+      statusCode: 404,
+      body: { message: 'Not Found' },
+    });
+  }
+
+  const uninstallTokenService = appContextService.getUninstallTokenService();
+  if (!uninstallTokenService) {
+    return response.customError({
+      statusCode: 500,
+      body: { message: 'Uninstall Token Service is unavailable.' },
+    });
+  }
+
+  try {
+    const uninstallTokens = await readUninstallTokens(
+      uninstallTokenService,
+      request.query.policyId
+    );
+
+    const { page = 1, perPage = 20 } = request.query;
+    const items = paginateTokens(uninstallTokens, perPage, page);
+
+    const body: GetUninstallTokensResponse = {
+      items,
+      total: Object.keys(uninstallTokens).length,
+      perPage,
+      page,
+    };
+
+    return response.ok({ body });
+  } catch {
+    return response.customError({
+      statusCode: 500,
+      body: { message: 'Failed to get uninstall tokens.' },
+    });
+  }
+};
+
+const readUninstallTokens = async (
+  uninstallTokenService: UninstallTokenServiceInterface,
+  policyIdFilter?: string
+) => {
+  let uninstallTokens: PolicyUninstallTokenMap;
+
+  if (policyIdFilter) {
+    const token = await uninstallTokenService.getTokenForPolicyId(policyIdFilter);
+
+    uninstallTokens = token ? { [policyIdFilter]: token } : {};
+  } else {
+    uninstallTokens = await uninstallTokenService.getAllTokens();
+  }
+
+  return uninstallTokens;
+};
+
+const paginateTokens = (
+  uninstallTokens: PolicyUninstallTokenMap,
+  perPage: number,
+  page: number
+) => {
+  const policyIds = Object.keys(uninstallTokens);
+  const items: PolicyUninstallTokenMap = {};
+
+  for (let i = perPage * (page - 1); i < Math.min(perPage * page, policyIds.length); i++) {
```

Comment:
this could potentially give different results if policies get added/deleted between page requests. I would be interested in hearing thoughts on how important this is.

I would prefer paging on the aggregation but it is higher effort so I'm leaning frontend paging given that policies are unlikely to exceed a thousand and even then those are rare cases.

Reply:
I think it's not a problem if the data changes between page requests - that's most likely expected, and probably better than being outdated for minutes in case somebody likes to switch between pages for a long time. also, and this is only an assumption that may be wrong, but I guess policies are not created so often, and searching by policyID will be the main use case.

yeah, with the current solution we don't win much with pagination: afaik reading the SOs from the 2nd time will be faster due to caching, and the user's browser needs to handle less data. but agree, paging on the aggregation would be better. I think the aggregation may change in the future with v2, so I'm leaning to keep this solution for now, and improve pagination later supported by the added integration tests - but frontend pagination could work, too, let's hear more opinions!

Reply:
I would avoid this API returning all the data in the response. 

About the pagination algorithm, is the service returning an object with the same policyId order always? My assumption is the service will return the data in a different order (since objects are not sorted) and this logic could end up returning different data for the same object and page/size, but correct me if I'm wrong or I'm missing something. Should the service return a sorted array (sorted by some field) instead of an object?

Reply:
I think it will probably make sense (and be more consistent with the rest of the Fleet API) to return an array of object uninstall token and the policy id and do the aggregation client side if needed, it will work better with pagination too.

What do you think?

Reply:
@dasansol92 makes a good point, this won't be sorted. @dasansol92 I'm curious as to your thoughts on why we should avoid returning all data in this case? I agree that it's generally not advisable but in this case where paging the aggregation is a bit more cumbersome, it seems like a good balance given that a thousand soft upper bound is a pretty small number and the objects are tiny.

@nchaulet can you help me understand what you mean by `do the aggregation client side`? does this mean to pass all uninstall token objects to frontend and write frontend logic to filter out latest for each policyId and also page on frontend?

another idea is that we can have the method additionally return saved object ids and then we can pass that into a new method that uses the ids to make a regular paginated request and returns a sorted paged array. this would allow us to rely on es for pagination.

Reply:
@joeypoon From a scalability standpoint, I would avoid the API returning too much data, specially when it has a dynamic length. Also I would avoid the UI doing extra work that can be done in the server side (filtering, pagination, etc). Even the objects are small, I would avoid that if possible.

Reply:
> @nchaulet can you help me understand what you mean by do the aggregation client side? does this mean to pass all uninstall token objects to frontend and write frontend logic to filter out latest for each policyId and also page on frontend?

sure what I had in mind is the API return an array in items instead of a map (it's more inline on how other Fleet APIs work)
```
items: [{ token, policyId, created_at },  ...]
```